Reenable cloning of serializable object for uapaot

ViktorHofer opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/18942

Local test verification build is still running. No failures yet. Please do not merge till it's finished.

danmosemsft commented:
What about these guys
```
C:\git\corefx\src\System.ObjectModel\tests\ObservableCollection\ObservableCollection_Serialization.cs:
   23:         [SkipOnTargetFramework(TargetFrameworkMonikers.UapAot, "Serialization still WIP in UapAot: https://github.com/dotnet/corefx/issues/18942")]

C:\git\corefx\src\System.ObjectModel\tests\ReadOnlyDictionary\ReadOnlyDictionary_SerializationTests.cs:
   22:         [SkipOnTargetFramework(TargetFrameworkMonikers.UapAot, "Serialization still WIP in UapAot: https://github.com/dotnet/corefx/issues/18942")]

C:\git\corefx\src\System.ObjectModel\tests\ReadOnlyObservableCollection\ReadOnlyObservableCollection_SerializationTests.cs:
   23:         [SkipOnTargetFramework(TargetFrameworkMonikers.UapAot, "Serialization still WIP in UapAot: https://github.com/dotnet/corefx/issues/18942")]
```

ViktorHofer commented:
Reenabled the tests (thanks @danmosemsft) and removed some test code which became unused after we removed serialization support for System.Security.Permissions.*.

@morganbr PTAL

ViktorHofer commented:
@morganbr System.Net.Primitives serialization tests are failing:

     System.Net.Primitives.Functional.Tests.SerializationTest.SerializeDeserialize_Roundtrip_EqualObjects(obj: somekey=
  somevalue) [FAIL]
        $BlockedFromReflection_2_091a32a8 : This object cannot be invoked because it was metadata-enabled for browsing
  only: 'System.Net.CookieVariant.value__' For more information, please visit  http://go.microsoft.com/fwlink/?LinkID=6
  16867

ViktorHofer commented:
@Petermarcu @joshfree The above error link `http://go.microsoft.com/fwlink/?LinkID=6` when running tests with TargetGroup=uapaot points to `https://www.amazon.com`. Do we have a new partnership I am not aware of? ;) I think we should fix the link to point to the respective doc page? Not sure whom to contact for that.

ViktorHofer commented:
@AtsushiKan Morgan is currently on holiday. He mentioned that it's strange that I need a rd.xml here. Could you please verify that? I had to add it as the System.Net.Primitives.CookieVariant enum wasn't reflectable (see error above). Thanks!

AtsushiKan commented:
Does all of System.Net.Primitives need rd.xml or just that one type (https://github.com/dotnet/corefx/blob/faefcc1653fe00c04e4156a9086e9b998e6214db/src/Microsoft.CSharp/src/Resources/Microsoft.CSharp.rd.xml#L92 for template of how to do a  single type.)

AtsushiKan commented on src/Common/tests/System/Runtime/Serialization/Formatters/BinaryFormatterHelpers.cs:
```diff
@@ -30,13 +26,6 @@ internal static T Clone<T>(T obj)
 
         internal static Lazy<T> Clone<T>(Lazy<T> lazy)
```
I think you can remove this overload entirely now. It was only needed for the .NETNative escape path.

ViktorHofer replied:
thanks

Petermarcu commented:
The error link works for me if I copy the whole id.

ViktorHofer commented:
Yes, sorry that was a mistake from my side!

ViktorHofer commented:
@AtsushiKan I'm not sure how many types need to be in the .rd.xml. Is there a trick how to find that out? I already tried with /p:ILCBuildType=chk but can't find any warnings or error messages. I tried just the one and also a few others but tests aren't passing...

AtsushiKan commented:
I know of no special tricks - it's rinse and repeat though usually there a MissingMetadataExceptions to sign post the way.

If it's going to be an endless treadmill, I don't think we need to block the PR on making the rd.xml more specific.

ViktorHofer commented:
@CIPop While working on this PR I found this code which doesn't seem correct. Shouldn't it be `'$(TargetGroup)' == 'uapaot'`? https://github.com/dotnet/corefx/blob/master/src/System.Net.Primitives/src/System.Net.Primitives.csproj#L190

ViktorHofer commented:
@CIPop and another question. why the asterisks here? https://github.com/dotnet/corefx/blob/master/src/System.Net.Primitives/src/Resources/System.Net.Primitives.rd.xml#L3

ViktorHofer commented:
@AtsushiKan @morganbr all errors should be fixed now. I refined to rd.xmls to only contain what's necessary.

ViktorHofer commented:
Thanks! Waiting for Morgans feedback :)

CIPop commented:
> Shouldn't it be '$(TargetGroup)' == 'uapaot'? Or could it be that uapaot also adds uap as the TargetGroup?
> why the asterisks here? 

@ViktorHofer good questions. I was following existing patterns for the *. I think you're right: `uap` must include `uapaot` otherwise the compilation for HttpListener would be broken.

@morganbr and @AtsushiKan might have more info.

ViktorHofer commented:
@CIPop let's discuss that here: https://github.com/dotnet/corefx/pull/22794

davidsh commented:
>@ViktorHofer good questions. I was following existing patterns for the *. I think you're right: uap must include uapaot otherwise the compilation for HttpListener would be broken.

There is no product source compilation for UAPAOT.  There is only a UAP compilation for PRODUCT sources.

UAPAOT mode is only for test sources if necessary since it invokes the .NET Native compiler after the test runner builds.

But in general, there is only a UAP binary in the NuGet packages.  .NET Native process happens after packages are built and only when a developer builds an app and then selects to use .NET Native on it.

ViktorHofer commented:
@dotnet-bot test OSX x64 Debug Build

@morganbr just a reminder to review the PR please

morganbr commented on src/System.Collections.Specialized/tests/Resources/System.Collections.Specialized.Tests.rd.xml:
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
+  <Library Name="System.Collections.Specialized.Tests">
+    <Type Name="System.Collections.Specialized.ListDictionary" Dynamic="Required All" />
```
You should be able to use BinaryFormatter degree instead of Dynamic for this (it should propagate to members).

morganbr commented on src/System.Collections/tests/Resources/System.Collections.Tests.rd.xml:
```diff
@@ -1,7 +1,15 @@
 <Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
-  <Library>
+  <Library Name="System.Collections.Tests">
+    <Assembly Name="*Application*" DoNotInline="true" DoNotOptimize="true" />
```
Using DoNotOptimize and DoNotInline looks concerning since we usually use those for codegen bugs. Why did you enable those?

ViktorHofer replied:
They were already there before my changes: https://github.com/dotnet/corefx/blob/master/src/System.Collections/tests/Resources/System.Collections.Tests.rd.xml

ViktorHofer replied:
Removing these causes some tests to fail (not serialization tests).

MichalStrehovsky replied:
> Removing these causes some tests to fail (not serialization tests).

@safern Where do we track the codegen bug? I can't find any mention of it in the pull request that added it.

I did a sweep of the codebase for unaccounted codegen workarounds (e.g. #22475) but I forgot about this sneaky trick. We really should stick to the `<ILCBuildType Condition="'$(TargetGroup)' == 'uap'">chk</ILCBuildType>` pattern for this.

safern replied:
> @safern Where do we track the codegen bug? I can't find any mention of it in the pull request that added it.
I did a sweep of the codebase for unaccounted codegen workarounds (e.g. #22475) but I forgot about this sneaky trick. We really should stick to the <ILCBuildType Condition="'$(TargetGroup)' == 'uap'">chk</ILCBuildType> pattern for this.

This bug is tracked in VSTS, bug number: 472947

Kirstie Lee from the codegen bug is working on getting that bug fixed. However we don't have a corefx tracking bug, but I am working with her to get it fixed. @ViktorHofer feel free to delete this file and add `<ILCBuildType>chk</ILCBuildType>` to the csproj. If you want to merge this ASAP let me know and I can take care of doing this update.

safern replied:
Also I added this rd.xml before we started using the ILCBuildType pattern. If @ViktorHofer wants to merge this now I can update it since I'm tracking this in my backlog.

ViktorHofer replied:
Thanks Santi for the info! Don't bother with it. I'm submitting a commit for it now.

MichalStrehovsky replied:
> This bug is tracked in VSTS, bug number: 472947

Oh, okay. The title of bug 472947 says it's for System.Data.SqlClient.Tests and this is System.Collections (plus that bug is only 3 days old, while this workaround has been here for a while) - that's why the confusion.

safern replied:
Thanks for taking care of it. Please do add a comment mentioning VSTS bug number 472947.

@MichalStrehovsky don't worry about cleaning this things up, I've been submitting the PRs and porting them to clean up all the tests projects that have ILCBuildType=chk and will do so for this one as well whenever the bug is fixed. I think this is the last test project with ILCBuildType=chk.

safern replied:
> Oh, okay. The title of bug 472947 says it's for System.Data.SqlClient.Tests and this is System.Collections (plus that bug is only 3 days old, while this workaround has been here for a while) - that's why the confusion.

Yup, sorry about that, I didn't open the bug they did in the codegen team and I was following up on email getting a consistent repro with them.

ViktorHofer replied:
>  think this is the last test project with ILCBuildType=chk.

Correct :)

morganbr commented on src/System.Collections/tests/Resources/System.Collections.Tests.rd.xml:
```diff
@@ -1,7 +1,15 @@
-  <Library>
+  <Library Name="System.Collections.Tests">
+    <Assembly Name="*Application*" DoNotInline="true" DoNotOptimize="true" />
+    <Assembly Name="System.Collections" BinaryFormatter="All" Dynamic="Required All" DoNotInline="true" DoNotOptimize="true" />
```
You shouldn't need both Dynamic and BinaryFormatter. For this test coverage, please prefer BinaryFormatter.

morganbr commented on src/System.Collections/tests/Resources/System.Collections.Tests.rd.xml:
```diff
@@ -1,7 +1,15 @@
     <Type Name="System.Collections.Generic.SortedList`2" Browse="Required Public" />
-    <Type Name="System.Collections.Generic.SortedDictionary`2" Browse="Required Public" />
+    <Type Name="System.Collections.Generic.SortedDictionary`2" Browse="Required Public">
+      <Field Name="_set" Dynamic="Required" />
```
Does it work to use BinaryFormatter on the type instead of Dynamic on the field?

morganbr commented on src/System.Collections/tests/Resources/System.Collections.Tests.rd.xml:
```diff
@@ -1,7 +1,15 @@
+    <Type Name="System.Collections.Generic.SortedDictionary`2" Browse="Required Public">
+      <Field Name="_set" Dynamic="Required" />
+    </Type>
+    <Type Name="System.Collections.Generic.SortedList`2" Dynamic="Required All" />
```
Same comment on preferring BinaryFormatter over Dynamic

morganbr commented on src/System.Net.Primitives/tests/FunctionalTests/Resources/System.Net.Primitives.Functional.Tests.rd.xml:
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
+  <Library Name="System.Net.Primitives.Functional.Tests">
+    <Type Name="System.Net.CookieVariant" BinaryFormatter="All" Dynamic="Required Public" />
```
You shouldn't need both BinaryFormatter and Dynamic

morganbr commented on src/System.Net.Primitives/tests/FunctionalTests/SerializationTest.cs:
```diff
@@ -13,7 +13,7 @@ public static class SerializationTest
         {
             yield return new object[] { new Cookie("somekey", "somevalue") };
-            yield return new object[] { new CookieCollection() { new Cookie("somekey", "somevalue") } };
+            // yield return new object[] { new CookieCollection() { new Cookie("somekey", "somevalue") } };
```
Why is this commented?

ViktorHofer commented:
Thanks for the feedback. The code paths where I'm using an rd.xml were failing after re-enabling the tests. I haven't investigated every case but most of them require reflecting on private fields.

ViktorHofer commented:
@morganbr PTAL

ViktorHofer commented:
@dotnet-bot test this please

davidsh commented on src/System.Net.Primitives/tests/FunctionalTests/Resources/System.Net.Primitives.Functional.Tests.rd.xml:
```diff
@@ -0,0 +1,7 @@
+    <Type Name="System.Net.CookieVariant" BinaryFormatter="All" />
+    <Type Name="System.Net.CookieCollection" BinaryFormatter="All" />
+  </Library>
+</Directives>
```
Why are we making the changes here in the test library rd.xml file?  Shouldn't his be in the product source library System.Net.Primitives?  What does this rd.xml do here?

If we need to expose information for these specific types (CookieVariant, CookieCollection) for BinaryFormatter to work, this rd.xml file should be in the product library and not the test library.  Otherwise, all developers need to add this to their own rd.xml file.  Or that is the real goal here?  I.e., we want to force developers to have to do this?

cc: @danmosemsft @weshaggard @safern @ViktorHofer

morganbr replied:
Short version: Our tests are harder to make work without rd.xml than customer code will be.

Long version: As with any reflection, the .NET Native compiler needs to know what to keep for BinaryFormatter. If it keeps all serializable types in all apps, that will add considerable size, especially for apps that don't use BinaryFormatter at all. Of course, we generally consider it a failure when customers have to add rd.xml, so we've added BinaryFormatter-specific analysis to the .NET Native toolchain that understands common patterns like
```
BinaryFormatter bf = new BinaryFormatter();
bf.Serialize(someObject);
```
That includes cases like calling through wrapper methods etc. However, these and other test cases look like:
```
        public static IEnumerable<object[]> SerializeDeserialize_Roundtrip_MemberData()
        {
            yield return new object[] { new Cookie("somekey", "somevalue") };
            yield return new object[] { new CookieCollection() { new Cookie("somekey", "somevalue") } };
        }
        [Theory]
        [MemberData(nameof(SerializeDeserialize_Roundtrip_MemberData))]
        public static void SerializeDeserialize_Roundtrip_EqualObjects(object obj)
        {
            Assert.Equal(obj, BinaryFormatterHelpers.Clone(obj));
        }
```

So there's not really a straight line between the inputs to Clone and the objects returned by  SerializeDeserialize_Roundtrip_MemberData. Since we mostly don't expect customers to build code that indirect in most cases, we're opting for rd.xml in the test.

davidsh replied:
That's a good explanation.  I think we should summarize that and add it as comments in the rd.xml file for these tests.  Otherwise, we won't remember why we added it.

safern commented on src/System.Collections/tests/System.Collections.Tests.csproj:
```diff
@@ -2,6 +2,8 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
+    <!-- Codegen bug, VSTS tracking number 472947 -->
+    <ILCBuildType Condition="'$(TargetGroup)' == 'uap'">chk</ILCBuildType>
```
This condition will never be true as this test project doesn't have a uap or uapaot configuration and it will cause a regression on the tests. Could you please remove the condition? 

https://github.com/dotnet/corefx/blob/master/src/System.Collections/tests/Configurations.props#L5-L6

ViktorHofer replied:
you are right. thanks!

ViktorHofer commented:
I will look through the code and talk with @morganbr about adding comments where necessary.
